includes/startup.inc

```diff
@@ -0,0 +1,260 @@
+      // value of [1 => 'value'] to cover this case.  If
+      // explode returns two items, the default value is ignored.
+      list($key, $value) = explode('=', $item, 2) + array(1 => '');
+      $env[$key] = $value;
+    }
+  }
+
+  return $env;
+}
+
+/**
+ * Checks the provided location and return the appropriate
+ * Drush wrapper or Drush launcher script, if found.
+ */
+function find_wrapper_or_launcher($location) {
+  // We look for a "Drush wrapper" script that might
+  // be stored in the root of a site.  If there is
+  // no wrapper script, then we look for the
+  // drush.launcher script in vendor/bin.  We try just a
+  // few of the most common locations; if the user relocates
+  // their vendor directory anywhere else, then they must
+  // use a wrapper script to locate it.  See the comment in
+  // 'examples/drush' for details.
+  //
+  // We are somewhat "loose" about whether we are looking
+  // for "drush" or "drush.launcher", because in old versions
+  // of Drush, the "drush launcher" was named "drush".
+  // Otherwise, there wouldn't be any point in looking for
+  // "drush.launcher" at the root, or "drush" in a vendor directory.
+  // We also allow users to rename their drush wrapper to
+  // 'drush.wrapper' to avoid conflicting with a directory named
+  // 'drush' at the site root.
+  $drush_locations = array(
+    "drush",
+    "vendor/bin/drush",
+    "../vendor/bin/drush",
+    "sites/all/vendor/bin/drush",
+  );
+
+  foreach ($drush_locations as $d) {
+    foreach (array('.launcher', '.wrapper', '') as $suffix) {
+      if (is_file("$location/$d$suffix")) {
+        return "$location/$d$suffix";
+      }
+    }
+  }
+  return "";
+}
+
+/**
+ * drush_startup is called once, by the Drush "finder"
+ * script -- the "drush" script at the Drush root.
+ * It finds the correct Drush "wrapper" or "launcher"
+ * script to use, and executes it with process replacement.
+ */
+function drush_startup($argv) {
+  $found_script = "";
+  $home = getenv("HOME");
+  $use_dir = "$home/.drush/use";
+
+  // Get the arguments for the command.  Shift off argv[0],
+  // which contains the name of this script.
+  $arguments = $argv;
+  array_shift($arguments);
+
+  //
+  // We need to do at least a partial parsing of the options,
+  // so that we can find --root / -r and --local
+  //
+  $VERBOSE=FALSE;
+  $LOCAL=FALSE;
+  $ROOT=FALSE;
+  $COMMAND=FALSE;
+  $ALIAS=FALSE;
+  $VAR=FALSE;
+
+  foreach ($arguments as $arg) {
+    // If a variable to set was indicated on the
+    // previous iteration, then set the value of
+    // the named variable (e.g. "ROOT") to "$arg".
+    if ($VAR) {
+      $$VAR = "$arg";
+      $VAR = FALSE;
+    }
+    else {
+      switch ($arg) {
+        case "--local":
+          $LOCAL = TRUE;
+          break;
+
+        case "-r":
+          $VAR = "ROOT";
+          break;
+
+        case "--verbose":
+        case "-v":
+          $VERBOSE = TRUE;
+          break;
```

Comment:
I've only included flags here that I use in the script.  I could put --debug in for future use, though.